Terraform Framework (#15)
gravitational/protoc-gen-terraform: README.md

# protoc-gen-terraform

protoc plugin to generate Terraform Framework schema definitions and getter/setter methods from gogo/protobuf .proto files.

# Installation

Install the generator binary.

```
go install github.com/gravitational/protoc-gen-terraform
```

# Usage

```
mkdir -p ./tfschema
protoc \
    -I$(teleport_dir)/api/types \
    -I$(teleport_dir)/vendor/github.com/gogo/protobuf \
    -I$(srcpath) \
    --plugin=./_build/protoc-gen-terraform \
    --terraform_out=types=types.UserV2+types.RoleV3,pkg=types:tfschema \
    types.proto
```

This command will generate `types_terraform.go` in `tfschema` folder. 

See [Makefile](Makefile) for details.

# Options

> Options could be set either using cli args or [YAML](example/teleport.yaml). Path to config file can be specified with `config` argument. Be advised that some options could be set in config file only.

Author: I guess, it is not needed anymore. Changed this to `test/config.yaml` instead.

# protoc-gen-terraform

protoc plugin to generate Terraform Framework schema definitions and getter/setter methods from gogo/protobuf .proto files.

# Installation

Install the generator binary.

```
go install github.com/gravitational/protoc-gen-terraform
```

# Usage

```
mkdir -p ./tfschema
protoc \
    -I$(teleport_dir)/api/types \
    -I$(teleport_dir)/vendor/github.com/gogo/protobuf \
    -I$(srcpath) \
    --plugin=./_build/protoc-gen-terraform \
    --terraform_out=types=types.UserV2+types.RoleV3,pkg=types:tfschema \

>     types.proto

Author: srcpath supposed to be go env GOPATH, replaced it here.

# protoc-gen-terraform

protoc plugin to generate Terraform Framework schema definitions and getter/setter methods from gogo/protobuf .proto files.

# Installation

Install the generator binary.

```
go install github.com/gravitational/protoc-gen-terraform
```

# Usage

Given that you have `gogo/protobuf` and `gravitational/teleport/api` in your $GOSRC dir:

```
mkdir -p ./tfschema
protoc \
    -I$(go env GOPATH)/src/github.com/gravitational/teleport/api/types \
    -I$(go env GOPATH)/src/github.com/gogo/protobuf \
    -I$(go env GOPATH)/src \
    --plugin=./build/protoc-gen-terraform \
    --terraform_out=types=RoleSpecV4,pkg=types:tfschema \
    types.proto
```

This command will generate `types_terraform.go` in `tfschema` folder. 

See [Makefile](Makefile) for details.

# Options

> Options could be set either using cli args or [YAML](test/config.yaml). Path to config file can be specified with `config` argument. Be advised that some options could be set in config file only.

Reviewer: ```suggestion
Options can be set using either CLI args or [YAML](test/config.yaml). The path to the config file can be specified with `config` argument. Be advised that some options can only be set via the config file.
```

```

## Excluding fields

Let's consider we have the following proto definition:

```proto
message Metadata {
	string ID = 1;
}

message User {
	Metadata Metadata = 1;
}

message AuthPreference {
	Metadata Metadata = 1;
}
```

Specify `exclude_fields` option:

```
exclude_fields=Metadata.ID+AuthPreference.Metadata.Name
```

In this case, `Metadata.ID` would be omitted for both `User` and `AuthPreference`, and `Metadata.Name` would be omitted for `AuthPreference` only. `User.Metadata.Name` won't be affected.

## Terraform Schema flags

You can specify `Required: true` (`required_fields`), `Computed: true` (`computed_fields`) and `Sensitive: true` (`sensitive_fields`) flags for your Terraform schema:

```
required_fields=Metadata.Name
```

You also can set list of `Validators` and `PlanModifiers` using configuration file:

```yaml
validators:
	"Metadata.Expires":
		- rfc3339TimeValidator

plan_modifiers:
	"Role.Options":
		- "github.com/hashicorp/terraform-plugin-framework/tfsdk.RequiresReplace()"
```

## UseStateForUnknown by default

The following setting:

```
use_state_for_unknown_by_default: true
```

will add `tfsdk.UseStateForUnknown()` PlanModifier to all computed fields.

> ## Inhjecting fields into schema

Reviewer: ```suggestion
## Injecting fields into schema
```
